runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -102,15 +102,21 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
       end_lnum = _end.line,
       end_col = line_byte_from_position(buf_lines, _end.line, _end.character, offset_encoding),
       severity = severity_lsp_to_vim(diagnostic.severity),
-      message = diagnostic.message
+      message = diagnostic.message,
+      lsp = {
```

Comment:
I had something like `user_data`  in mind (Similar to how it is used in `complete-items`) and nest it further - to avoid making `lsp` a first-class diagnostic item property. 

But no strong opinion.

Reply:
Also no strong opinion. Do we anticipate additional things populating this field? I don't know of any off the top of my head, but additional table nesting doesn't heart (apart from code beauty)

Reply:
I don't anticipate needing to attach any other kind of non-LSP related data here, but of course I can't see the future and this would save us a refactor if that ever was the case. So SGTM.